[Core] [Doc] Add OSS Document for Task Events (#56203)
ray-project/ray: doc/source/ray-observability/user-guides/ray-event-export.rst

.. _ray-event-export:

Ray Event Export
================

Ray support exporting structured events to a configured HTTP endpoint. Each worker node
sends events to the configured HTTP endpoint through a HTTP POST request.

Starting from 2.49, Ray supports exporting task events. Ray plans to support other 
event types, including actor events, node events, job events, etc., in the future.  

.. note:: 
    Ray event export feature is still in alpha version. The way to configure event 
    reporting and the format of the events are subject to change.

> Enable event reporting

Reviewer: What do you think about having a section briefly describing the architecture of the aggregator agent and listing some env vars to configure the buffer?

Author: I think it is a good idea to describe the architecture. I'll add one section for that. 

At the same time, I'm a little bit hesitate to expose the buffer configuration now. 

The reason are that: 
(1) env vars might not be a good way for the users to configure Ray and I'm planning to revamp the section after we check in the code to add the command line parameter & ray.init() parameter to configure the event export destination; 
(2) we've already exposed a lot of API to the users today and we should only expose new ones if they are absolutely necessary. So without user feedback, I think we should default to expose as less APIs as possible.

----------------------

> To enable event reporting, you need to set the ``RAY_enable_core_worker_ray_event_to_aggregator`` environment 

Reviewer 2: ```suggestion
To enable event reporting, you need to set the ``RAY_enable_core_worker_ray_event_to_aggregator`` environment 
```

Should the env var be something like `RAY_enable_ray_event_export` given events can be from raylet, gcs etc as well?

Author: From my perspective, this env var will only be temporarily used before we add the public API to ray start scrip & ray.init() and before we migrate the GCS path. With the 2 changes done, we can remove the environment variable and only use the values configured from the public API to determine whether to send the event to GCS. We will revamp the doc then as well.

variable to ``1`` when starting each Ray worker node.

> To set the target HTTP end point, you need to set the ``RAY_events_export_addr`` 

Reviewer 2: ```suggestion
To set the target HTTP endpoint, you need to set the ``RAY_events_export_addr`` 
```

> environment variable. The value should be a valid HTTP URL with the scheme of ``http``.

Reviewer 2: can it be https?

Author: Currently it only supports `http`. I think we can add `https` support later.

Event format
------------

Events are JSON objects in the POST request body.

All events contains the same base fields and different event specific fields. 

> See ``src/ray/protobuf/events_base_event.proto`` for the base fields.

Reviewer 2: can you create a link to the file?

            "requiredResources":{
                "CPU":1.0
            },
            "runtimeEnvInfo":{
                "serializedRuntimeEnv":"{}",
                "runtimeEnvConfig":{
                    "setupTimeoutSeconds":600,
                    "eagerInstall":true,
                    "logFiles":[
                    
                    ]
                }
            },
            "jobId":"AQAAAA==",
            "parentTaskId":"//////////////////////////8BAAAA",
            "placementGroupId":"////////////////////////",
            "taskAttempt":0,
            "taskType":"NORMAL_TASK",
            "language":"PYTHON",
            "refIds":{
                
            }
        },
        "message":""
    }

    // task execution event
    {
        "eventId":"vkIaAHlQC5KoppGosqs2kBq5k2WzsAAbawDDbQ==",
        "sourceType":"CORE_WORKER",
        "eventType":"TASK_EXECUTION_EVENT",
        "timestamp":"2025-09-03T18:52:14.469074Z",
        "severity":"INFO",
        "sessionName":"session_2025-09-03_11-52-12_635210_85618",
        "taskExecutionEvent":{
            "taskId":"yO9FzNARJXH///////////////8BAAAA",
            "taskState":{

>                 "2":"2025-09-03T18:52:14.467402Z",

Author: They map to the TaskStatus in `common.proto`

                "1":"2025-09-03T18:52:14.467290Z",

>                 "5":"2025-09-03T18:52:14.469074Z"

Reviewer 2: oh, we show 1,2,3 here instead of the state string?

.. _ray-event-export:

Ray Event Export
================

> Ray supports exporting structured events to a configured HTTP endpoint. Each worker node

Author: Good point. Will do.

sends events to the configured HTTP endpoint through a HTTP POST request.

Starting from 2.49, Ray supports exporting task events. Ray plans to support other 
event types, including actor events, node events, job events, etc., in the future.  

.. note:: 
    Ray event export feature is still in alpha version. The way to configure event 
    reporting and the format of the events are subject to change.

Enable event reporting
----------------------
To enable event reporting, you need to set the ``RAY_enable_core_worker_ray_event_to_aggregator`` environment 
variable to ``1`` when starting each Ray worker node.

To set the target HTTP endpoint, you need to set the ``RAY_events_export_addr`` 

> environment variable. The value should be a valid HTTP URL with the scheme of ``http``.

Author: I'll probably change `the scheme of http` to `the http:// URL scheme`, if that sounds better.

Event format
------------

Events are JSON objects in the POST request body.

All events contains the same base fields and different event specific fields. 
See `src/ray/protobuf/public/events_base_event.proto <https://github.com/ray-project/ray/blob/master/src/ray/protobuf/public/events_base_event.proto>`_ for the base fields.

Task events
^^^^^^^^^^^

For each task, Ray exports two types of events: Task Definition Event and Task Execution Event.

* Task Definition Event generated once per task attempt. It contains the metadata of the task. 
  See `src/ray/protobuf/public/events_task_definition_event.proto <https://github.com/ray-project/ray/blob/master/src/ray/protobuf/public/events_task_definition_event.proto>`_ 
  and `src/ray/protobuf/public/events_actor_task_definition_event.proto <https://github.com/ray-project/ray/blob/master/src/ray/protobuf/public/events_actor_task_definition_event.proto>`_ for the event format for normal tasks 
  and actor tasks respectively.

> * Task Execution Event contains the task state transition information and metadata that 
>   generated during the task execution. 

Reviewer 3: ```suggestion
* Task Execution Events contain task state transition information and metadata 
  generated during task execution. 
```

  See `src/ray/protobuf/public/events_task_execution_event.proto <https://github.com/ray-project/ray/blob/master/src/ray/protobuf/public/events_task_execution_event.proto>`_ for the event format.

> An example of the task events is as follows:

Reviewer 3: ```suggestion
An example of a Task Definition Event followed by a Task Execution Event:
```
? I don't know if this is right

        "sourceType":"CORE_WORKER",
        "eventType":"TASK_EXECUTION_EVENT",
        "timestamp":"2025-09-03T18:52:14.469074Z",
        "severity":"INFO",
        "sessionName":"session_2025-09-03_11-52-12_635210_85618",
        "taskExecutionEvent":{
            "taskId":"yO9FzNARJXH///////////////8BAAAA",
            "taskState":{
                // key is the integer value of TaskStatus enum in common.proto at
                // https://github.com/ray-project/ray/blob/master/src/ray/protobuf/common.proto
                "2":"2025-09-03T18:52:14.467402Z", // PENDING_NODE_ASSIGNMENT
                "1":"2025-09-03T18:52:14.467290Z", // PENDING_ARGS_AVAIL
                "5":"2025-09-03T18:52:14.469074Z" // SUBMITTED_TO_WORKER
            },
            "nodeId":"ZvxTI6x9dlMFqMlIHErJpg5UEGK1INsKhW2zyg==",
            "workerId":"hMybCNYIFi+/yInYYhdc+qH8yMF65j/8+uCTmw==",
            "jobId":"AQAAAA==",
            "taskAttempt":0,
            "workerPid":0
        },
        "message":""
    }

High-level Architecture
-----------------------

The following diagram shows the high-level architecture of Ray event export.

.. image:: ../images/ray-event-export.png

> Ray introduces a new aggregator agent on the head node and each worker nodes. All the 
> Ray components send events to the aggregator agent through gRPC. The aggregator agent 
> collects the events and sends them to the configured HTTP endpoint. 

Reviewer 3: ```suggestion
All Ray components send events to an aggregator agent through gRPC. There is an aggregator
agent on each node. The aggregator agent collects all events on that node and sends the
events to the configured HTTP endpoint. 
```
